vyper/parser/parser_utils.py

```diff
@@ -747,6 +750,38 @@ def visit_UnaryOp(self, node):
             return node
 
 
+class UnmatchedReturnChecker(ast.NodeVisitor):
+    """
+    Make sure all return statement are balanced
+    (both branches of if statement should have returns statements).
+    """
+
+    def visit_FunctionDef(self, node: ast.FunctionDef) -> None:
+        self.handle_primary_function_def(node)
+
+    def handle_primary_function_def(self,  node: ast.FunctionDef) -> None:
+        if node.returns and not self.return_check(node.body):
+            raise StructureException(
+                f'Missing or Unmatched return statements in function "{node.name}". '
+                'All control flow statements (like if) need balanced return statements.',
+                node
+            )
+
+    def return_check(self, node: Union[ast.AST, List[Any]]) -> bool:
+        if isinstance(node, ast.Return):
```

Comment:
As mentioned on gitter, this sould be `isinstance(node, ast.Return) and isinstance(node, ast.Raise)` to allow:

```
if 1 == 1 :
  return 1
else :
  raise "unreachable"
```
Which should be allowed.

Reply:
y tho? isn't this dead code?

Reply:
here's another example
```
if x == 1 : 
  return 1
elif y == 2 : 
  return 3
raise "fail"
```
`raise` functions like `return` (except for setting the success flag to 0 for the caller). so it removes the requirement for the programmer to insert an unreachable dummy return after a raise. and maybe we should do the same for `selfdestruct`.

Reply:
No I meant `1 == 1` lol

Reply:
@fubuloubu dead code is beyond the scope of this PR. Please defer discussion to #1385 rather ;)